apps/condo/domains/acquiring/schema/GeneratePaymentLinkService.js

```diff
@@ -0,0 +1,270 @@
+        variable: ['data', 'receiptData', 'paymentYear'],
+        code: BAD_USER_INPUT,
+        type: RECEIPT_HAVE_INVALID_PAYMENT_YEAR_VALUE,
+        message: 'Cannot generate payment link with invalid "paymentYear" value',
+    },
+}
+
+const GeneratePaymentLinkService = new GQLCustomSchema('GeneratePaymentLinkService', {
+    types: [
+        {
+            access: true,
+            type: 'input GeneratePaymentLinkCallbacksInput { successUrl: String!, failureUrl: String! }',
+        },
+        {
+            access: true,
+            type: 'input GeneratePaymentLinkReceiptDataInput { currencyCode: String!, amount: String!, periodYear: Int!, periodMonth: Int!, accountNumber: String! }',
+        },
+        {
+            access: true,
+            type: 'input GeneratePaymentLinkInput { dv: Int!, sender: SenderFieldInput!, receipt: BillingReceiptWhereUniqueInput, receiptData: GeneratePaymentLinkReceiptDataInput, acquiringIntegrationContext: AcquiringIntegrationContextWhereUniqueInput!, callbacks: GeneratePaymentLinkCallbacksInput! }',
+        },
+        {
+            access: true,
+            type: 'type GeneratePaymentLinkOutput { dv: Int!, paymentUrl: String! }',
+        },
+    ],
+
+    queries: [
+        {
+            access: access.canGeneratePaymentLink,
+            schema: 'generatePaymentLink(data: GeneratePaymentLinkInput!): GeneratePaymentLinkOutput',
+            resolver: async (parent, args, context) => {
+                // wrap validator function to the current call context
+                const { data } = args
+                const {
+                    receipt,
+                    receiptData,
+                    acquiringIntegrationContext,
+                    callbacks: { successUrl, failureUrl },
+                } = data
+
+                // Stage 0: Check if sender input is valid
+                checkDvAndSender(data, errors.DV_VERSION_MISMATCH, errors.WRONG_SENDER_FORMAT, context)
+
+                // Stage 1: get acquiring context & integration
+                const acquiringContext = await getById('AcquiringIntegrationContext', acquiringIntegrationContext.id)
+
+                if (acquiringContext.deletedAt) {
+                    throw new GQLError(errors.ACQUIRING_INTEGRATION_CONTEXT_IS_DELETED, context)
+                }
+
+                // Stage 2: generate links
+                // two cases expected here
+                // first one: receipt.id was provided - a regular payment flow expected
+                // second one: receipt.id is empty, but receiptData are provided - a virtual receipt payment flow expected
+                const paymentLinkBaseUrl = new URL(`${conf.SERVER_URL}${PAYMENT_LINK_PATH}`)
+
+                // set common QP
+                paymentLinkBaseUrl.searchParams.set('aic', acquiringIntegrationContext.id)
+                paymentLinkBaseUrl.searchParams.set('su', successUrl)
+                paymentLinkBaseUrl.searchParams.set('fu', failureUrl)
```

Comment:
`aic` `su` and `fu` are very confusing.. Maybe move them to consts? Also, what is `QP` ?

Reply:
Ok, will do that. QP means query parameters.